crates/deltalake-core/src/logstore/mod.rs

```diff
@@ -0,0 +1,340 @@
+//! Delta log store.
+use futures::StreamExt;
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{
+    de::{Error, SeqAccess, Visitor},
+    ser::SerializeSeq,
+    Deserialize, Serialize,
+};
+use std::{cmp::max, collections::HashMap, sync::Arc};
+use url::Url;
+
+use crate::{
+    errors::DeltaResult,
+    operations::transaction::TransactionError,
+    protocol::{get_last_checkpoint, ProtocolError},
+    storage::{commit_uri_from_version, config::StorageOptions},
+    DeltaTableError,
+};
+use bytes::Bytes;
+use log::debug;
+use object_store::{
+    path::Path, Error as ObjectStoreError, ObjectStore, Result as ObjectStoreResult,
+};
+
+#[cfg(feature = "datafusion")]
+use datafusion::datasource::object_store::ObjectStoreUrl;
+
+pub mod default_logstore;
+
+/// Sharable reference to [`LogStore`]
+pub type LogStoreRef = Arc<dyn LogStore>;
+
+lazy_static! {
+    static ref DELTA_LOG_PATH: Path = Path::from("_delta_log");
+}
+
+/// Configuration parameters for a log store
+#[derive(Debug, Clone)]
+pub struct LogStoreConfig {
+    /// url corresponding to the storage location.
+    pub location: Url,
+    /// Options used for configuring backend storage
+    pub options: StorageOptions,
+}
+
+/// Trait for critical operations required to read and write commit entries in Delta logs.
+///
+/// The correctness is predicated on the atomicity and durability guarantees of
+/// the implementation of this interface. Specifically,
+///
+/// - Atomic visibility: Any commit created via `write_commit_entry` must become visible atomically.
+/// - Mutual exclusion: Only one writer must be able to create a commit for a specific version.
+/// - Consistent listing: Once a commit entry for version `v` has been written, any future call to
+///   `get_latest_version` must return a version >= `v`, i.e. the underlying file system entry must
+///   become visible immediately.
+#[async_trait::async_trait]
+pub trait LogStore: Sync + Send {
+    /// Read data for commit entry with the given version.
+    /// TODO: return the actual commit data, i.e. Vec<Action>, instead?
```

Comment:
eventually I think yes. Right now this is actually better as I have some follow up work in flight proting stuff from kernel, that may impact this as well..

Reply:
IIUC I'll keep it at `-> DeltaResult<Bytes>` for now, resolving.